src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -197,6 +180,85 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageController<'a, C, T> {
+                let mut new_read_capability = policy.frontier(collection.write_frontier.frontier());
+
+                if <_ as timely::order::PartialOrder>::less_equal(
+                    &collection.implied_capability,
+                    &new_read_capability,
+                ) {
+                    let mut update = ChangeBatch::new();
+                    update.extend(new_read_capability.iter().map(|time| (time.clone(), 1)));
+                    std::mem::swap(&mut collection.implied_capability, &mut new_read_capability);
+                    update.extend(new_read_capability.iter().map(|time| (time.clone(), -1)));
+                    if !update.is_empty() {
+                        read_capability_changes.insert(id, update);
+                    }
+                }
+
+                collection.read_policy = policy;
+            } else {
+                tracing::error!("Reference to unregistered id: {:?}", id);
+            }
+        }
+        if !read_capability_changes.is_empty() {
+            self.update_read_capabilities(&mut read_capability_changes)
+                .await;
+        }
+    }
+
+    /// Assigns a read policy to specific identifiers.
+    ///
+    /// The policies are assigned in the order presented, and repeated identifiers should
+    /// conclude with the last policy. Changing a policy will immediately downgrade the read
+    /// capability if appropriate, but it will not "recover" the read capability if the prior
+    /// capability is already ahead of it.
+    ///
+    /// Identifiers not present in `policies` retain their existing read policies.
+    ///
+    /// This method is not `async` (yay) but will crash if it needs to transmit changes (boo).
+    pub fn set_read_policy_sync(&mut self, policies: Vec<(GlobalId, ReadPolicy<T>)>) {
+        let mut read_capability_changes = BTreeMap::default();
+        for (id, policy) in policies.into_iter() {
+            if let Ok(collection) = self.collection_mut(id) {
+                let mut new_read_capability = policy.frontier(collection.write_frontier.frontier());
+
+                if <_ as timely::order::PartialOrder>::less_equal(
+                    &collection.implied_capability,
+                    &new_read_capability,
+                ) {
+                    let mut update = ChangeBatch::new();
+                    update.extend(new_read_capability.iter().map(|time| (time.clone(), 1)));
+                    std::mem::swap(&mut collection.implied_capability, &mut new_read_capability);
+                    update.extend(new_read_capability.iter().map(|time| (time.clone(), -1)));
+                    if !update.is_empty() {
+                        read_capability_changes.insert(id, update);
+                    }
+                }
+
+                collection.read_policy = policy;
+            } else {
+                tracing::error!("Reference to unregistered id: {:?}", id);
+            }
+        }
```

Comment:
This should be extracted and shared with the async implementation, if nothing else.